modules/assim.sequential/R/sda.enkf_refactored.R

```diff
@@ -0,0 +1,393 @@
+#' 
+#' @description State Variable Data Assimilation: Ensemble Kalman Filter and Generalized ensemble kalman file=ter
+#' 
+#' @return NONE
+#' @import nimble
+#' @export
+#' 
+
+sda.enkf <- function(settings, obs.mean, obs.cov, Q = NULL, restart=F, 
+                     control=list(trace=T,
+                                  interactivePlot=T,
+                                  TimeseriesPlot=T,
+                                  BiasPlot=F,
+                                  plot.title=NULL,
+                                  debug=FALSE),...) {
+  #------------- Some important variables  
+  # muf/Pf - forecast mean and covariance
+  # Y/R    - Observed data and covariance
+  # mu.a/Pa  - after analysis - new mean and covariance
+  # nt is the length of observed  
+  # When processvar == FALSE it means we are doing EnKF and when it's TRUE Generalized Ensemble Filter
+  # Generalized Ensemble Filter NEEDS process variance to avoid filter divergence and it does not
+  # have analytical solution - needs MCMC
+  # X stores IC of state variables and then collects state variables in each loop
+  # Y stores the observed mean
+  # Assimilation is done for start:end setting assimilation section - if it's a continuation of another sda (restart)
+  # then start date in the second xml should be the same as the first pecan xml
+  # Models that they wanna be added for SDA their read_restart needs to be in a certain format. look into read_restart_SIPNET
+  #-------------------------------------------------------------------------------
+  
+  ymd_hms <- lubridate::ymd_hms
+  hms     <- lubridate::hms
+  second  <- lubridate::second
+  ###-------------------------------------------------------------------###
+  ### read settings                                                     ###
+  ###-------------------------------------------------------------------###
+  adjustment <- settings$state.data.assimilation$adjustment
+  model      <- settings$model$type
+  write      <- settings$database$bety$write
+  defaults   <- settings$pfts
+  outdir     <- settings$modeloutdir # currently model runs locally, this will change if remote is enabled
+  rundir     <- settings$host$rundir
+  host       <- settings$host
+  forecast.time.step <- settings$state.data.assimilation$forecast.time.step  #idea for later generalizing
+  nens       <- as.numeric(settings$ensemble$size)
+  processvar <- settings$state.data.assimilation$process.variance
+  var.names <- sapply(settings$state.data.assimilation$state.variable, '[[', "variable.name")
+  names(var.names) <- NULL
+  #filtering obs data based on years specifited in setting > state.data.assimilation
+  assimyears <- year(settings$state.data.assimilation$start.date) : year(settings$state.data.assimilation$end.date) # years that assimilations will be done for - obs will be subsetted based on this
```

Comment:
would you like to `lubridate::year` these (and the other two lines below)? or extract `year <- lubridate::year` above like the others?